plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Game.hs

```diff
@@ -1,68 +1,130 @@
+newtype ClearString = ClearString ByteString
 
 PlutusTx.makeLift ''ClearString
 
 correctGuess :: HashedString -> ClearString -> Bool
-correctGuess (HashedString actual) (ClearString guess') = actual == (sha2_256 guess')
+correctGuess (HashedString actual) (ClearString guess') = actual == sha2_256 guess'
 
+-- | The validator (datascript -> redeemer -> PendingTx -> Bool)
 validateGuess :: HashedString -> ClearString -> PendingTx -> Bool
 validateGuess dataScript redeemerScript _ = correctGuess dataScript redeemerScript
 
 gameValidator :: ValidatorScript
 gameValidator =
-    ValidatorScript ($$(Ledger.compileScript [|| validateGuess ||]))
+    Ledger.ValidatorScript ($$(Ledger.compileScript [|| validateGuess ||]))
 
 gameDataScript :: String -> DataScript
 gameDataScript =
-    DataScript . Ledger.lifted . HashedString . plcSHA2_256 . C.pack
+    Ledger.DataScript . Ledger.lifted . HashedString . Ledger.plcSHA2_256 . C.pack
 
 gameRedeemerScript :: String -> RedeemerScript
 gameRedeemerScript =
-    RedeemerScript . Ledger.lifted . ClearString . C.pack
+    Ledger.RedeemerScript . Ledger.lifted . ClearString . C.pack
 
 gameAddress :: Address
 gameAddress = Ledger.scriptAddress gameValidator
 
-lock :: (WalletAPI m, WalletDiagnostics m) => String -> Value -> m ()
-lock word vl = do
-    let ds = gameDataScript word
-    payToScript_ defaultSlotRange gameAddress vl ds
+-- | Parameters for the "lock" endpoint
+data LockParams = LockParams
+    { secretWord :: String
+    , amount     :: Ada
+    }
+    deriving stock (Prelude.Eq, Prelude.Ord, Prelude.Show, Generic)
+    deriving anyclass (Aeson.FromJSON, Aeson.ToJSON)
+
+--  | Parameters for the "guess" endpoint
+newtype GuessParams = GuessParams
+    { guessWord :: String
+    }
+    deriving stock (Prelude.Eq, Prelude.Ord, Prelude.Show, Generic)
+    deriving newtype (Aeson.FromJSON, Aeson.ToJSON)
+
+guess :: ContractActions r => Contract r ()
+guess = do
+    st <- nextTransactionAt gameAddress
+    let mp = AM.fromTxOutputs st
+    GuessParams theGuess <- endpoint "guess"
+    let
+        txOutputs  = Map.toList . fromMaybe Map.empty $ mp ^. at gameAddress
+        redeemer   = gameRedeemerScript theGuess
+        inp        = (\o -> Ledger.scriptTxIn (fst o) gameValidator redeemer) <$> txOutputs
+        tx         = unbalancedTx inp []
+    void (writeTx tx)
+
+lock :: ContractActions r => Contract r ()
+lock = do
+    LockParams secret amt <- endpoint "lock"
+    let
+        vl         = Ada.toValue amt
+        dataScript = gameDataScript secret
+        output     = Ledger.TxOutOf gameAddress vl (Ledger.PayToScript dataScript)
+        tx         = unbalancedTx [] [output]
+    void (writeTx tx)
+
+game :: ContractActions r => Contract r ()
+game = guess <|> lock
 
-guess :: (WalletAPI m, WalletDiagnostics m) => String -> m ()
-guess word = do
-    let redeemer = gameRedeemerScript word
-    collectFromScript defaultSlotRange gameValidator redeemer
+lockTrace
```

Comment:
Should these be in the test file?

Reply:
I think it's good to have one or two traces with the contract code, to demonstrate how the contract works (happy path) and also because the traces are useful for `Language.Plutus.Contract.App.runWithTraces`